Use async InstanceTracker in DeploymentActor

janisz opened the pull request:
Related to #3031 #4693

review-ninja commented:
<a href="https://app.review.ninja/mesosphere/marathon/pull/4788" target="_blank"><img src="https://app.review.ninja/10941409/pull/4788/badge" alt="ReviewNinja"/>

jasongilanfarr commented:
ok to test

jasongilanfarr commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala:
```diff
@@ -145,40 +143,41 @@ private class DeploymentActor(
 
-    killTasksIfNeeded.flatMap(_ => startTasksIfNeeded)
+    val futureRunningInstances = instanceTracker.specInstances(runnableSpec.id).map(_.filter(_.state.condition.isActive))
+    futureRunningInstances.flatMap({ runningInstances =>
```
"({" is redundant, please just use "{".

Also, like the other PRs outstanding, please prefer async/await over monadic methods on Futures.

janisz replied:
Fixed. Will fix this in other PR. Please check if this is good.

zen-dog commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala:
```diff
@@ -145,41 +144,52 @@ private class DeploymentActor(
+      val runningInstances = instances.filter(_.state.condition.isActive)
+      val ScalingProposition(tasksToKill, tasksToStart) = ScalingProposition.propose(
+        runningInstances, toKill, killToMeetConstraints, scaleTo, runnableSpec.killSelection)
+      await {
```
This `await` will wait for 2. future `startTasksIfNeeded` where `killTasksIfNeeded` may end afterwards. This is pretty much the opposite of what the original code was doing: wait for kills, start afterwards.

janisz commented:
Fixed issues and rebase

zen-dog commented on src/main/scala/mesosphere/marathon/upgrade/DeploymentActor.scala:
```diff
@@ -136,47 +137,51 @@ private class DeploymentActor(
-    }
 
-    killTasksIfNeeded.flatMap(_ => startTasksIfNeeded)
+      await(killTasksIfNeeded.flatMap(_ => startTasksIfNeeded))
```
I'm don't mind future monads but I'm not a big fan of mixing them with await/async. Let's just stick with async/await.

janisz replied:
Fixed

jdef commented:
failing integration test. related to this change?
```
Error Message

assertion failed: timeout (3 seconds) during expectMsgClass waiting for class mesosphere.marathon.MarathonSchedulerActor$DeploymentStarted
Stacktrace

      java.lang.AssertionError: assertion failed: timeout (3 seconds) during expectMsgClass waiting for class mesosphere.marathon.MarathonSchedulerActor$DeploymentStarted
      at scala.Predef$.assert(Predef.scala:170)
      at akka.testkit.TestKitBase$class.expectMsgClass_internal(TestKit.scala:472)
      at akka.testkit.TestKitBase$class.expectMsgType(TestKit.scala:445)
      at mesosphere.marathon.MarathonSchedulerActorTest.expectMsgType(MarathonSchedulerActorTest.scala:37)
      at mesosphere.marathon.MarathonSchedulerActorTest$$anonfun$17.apply(MarathonSchedulerActorTest.scala:487)
      at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
      at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
      at org.scalatest.Transformer.apply(Transformer.scala:22)
      at org.scalatest.Transformer.apply(Transformer.scala:20)
      at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)
      at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
      at mesosphere.marathon.MarathonSchedulerActorTest.withFixture(MarathonSchedulerActorTest.scala:37)
      at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:183)
      at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:196)
      at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:196)
      at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
      at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:196)
      at mesosphere.marathon.MarathonSchedulerActorTest.org$scalatest$BeforeAndAfter$$super$runTest(MarathonSchedulerActorTest.scala:37)
      at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:203)
      at mesosphere.marathon.MarathonSchedulerActorTest.runTest(MarathonSchedulerActorTest.scala:37)
      at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:229)
      at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:229)
      at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
      at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
      at scala.collection.immutable.List.foreach(List.scala:381)
      at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
      at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
      at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
      at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:229)
      at mesosphere.marathon.MarathonSchedulerActorTest.runTests(MarathonSchedulerActorTest.scala:37)
      at org.scalatest.Suite$class.run(Suite.scala:1147)
```

janisz commented:
I'll check this.

jeschkies commented:
@janisz thanks. Did you have time to look into the failing tests? Unfortunately I'm from tomorrow on so I won't have the time this year :/

janisz commented:
@jeschkies I fixed fix for https://github.com/mesosphere/marathon/pull/4788#discussion_r92003957 and rebased on current master. Rerun all tests 3 times and all was green. I'll run mentioned test in a loop to check if it's flaky.

zen-dog commented:
@janisz I've taken a look at the flaky `MarathonSchedulerActorTest` and actually found a problem with the `DeploymentManager`. This is being fixed here: https://phabricator.mesosphere.com/D378 Please rebase your branch after the fix lands on master. 
`AppsResourceTest` also failed but I couldn't make it fail locally.

zen-dog commented:
I'm not sure why this PR got closed.

zen-dog commented:
@janisz we've landed a few fixes for flaky tests. would you rebase please?

janisz commented:
@zen-dog rebased but it looks like master is unstable. AppStartActorTest is failing on my local

zen-dog commented:
@janisz sure you got the last version? I fixed the flaky `AppStartActorTest` in this commit: https://github.com/mesosphere/marathon/commit/24845d677749c7a3194f3936e263bdb58940748c
